### DISCUSSION TITLE
How many levels does support BFV/BGV?

### DISCUSSION THREAD

**Discussion by carblu** [tuneinsight/lattigo#516] [Q&A] [answered]
I am running some simple experiments. In BFV/BGV, I was trying to use as many levels as possible. It seems that loqQ can be formed by at most 32 sizes of primes. 

I'm conducting basic experiments with the BFV and BGV encryption schemes. My goal is to utilize as many levels as possible within these schemes. Nevertheless, I've encountered a limitation with loqQ, which seems to be restricted to a maximum of 32 prime sizes. For instance, the following parameters do not work. 

bgv.ParametersLiteral{
   LogN:            16,          
   LogQ:            []int{56, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30}, 
   LogP:            []int{60, 60, 60, 60, 60}, 
   PlaintextModulus: 0xc0001,       
}           

I can encrypt and decrypt, but when decoding the plaintext obtained as a result of the decryption, I get the runtime error "index out of range [32] with length 32" raised by ring.reconstructRNS.

**Comment by Pro7ech** [comment, marked as answer]
Indeed there is a hard-coded `[32]uint64` array in the `ring.BasisExtender` and `ring.Decomposer`. I think allocating parameters with more than 32 primes should return an error to avoid a panic at runtime, I encourage you to open an issue about it.

If you really want, you can modify this hard-coded value to something larger and it should work.

In theory over 32 primes should not be needed because the higher the level is the more expensive homomorphic operations are, especially key-switching, and it becomes more efficient to instead have a smaller starting level and use bootstrapping. Unfortunately Lattigo does not support bootstrapping for BFV/BGV, it only does for CKKS.

**Reply by carblu** [reply]
Thanks for the quick reply. I was simply curious about the "theoretical" maximum number of levels, pushing the implementation to its limits. I fully agree that a smaller L with bootstrapping is the most effective strategy. Anyway, I will open an issue about it.
